ciphey/basemods/Decoders/brainfuck.py

```diff
@@ -0,0 +1,172 @@
+            * Includes at least one putchar instruction (".")
+            * Does not contain anything but the main 7 instructions,
+                (excluding ",") and whitespace (tabs, spaces, and newlines)
+
+        Details:
+            * This implementation wraps the memory pointer for ">" and "<"
+            * It is time-limited to 60 seconds, to prevent hangups
+            * The program starts with 100 memory cells, chosen arbitrarily
+        """
+
+        result = ""
+        memory = [0] * 100
+        codeptr, memptr = 0, 0      # instruction pointer, stack pointer
+        timelimit = 60              # timeout, in seconds
+
+        bracemap, isbf = self.bracemap_and_check(ctext)
+
+        # this doesn't appear to be a brainfuck program we can decode
+        if not isbf:
+            return None
+
+        # get start time
+        start = time.time()
+
+        while codeptr < len(ctext):
+
+            current = time.time()
+
+            # arbitrarily quit if we've been running for over a minute
+            if current - start > timelimit:
+                return None
+
+            cmd = ctext[codeptr]
+
+            if cmd == "+":
+                if memory[memptr] < 255:
+                    memory[memptr] = memory[memptr] + 1
+                else:
+                    memory[memptr] = 0
+
+            elif cmd == "-":
+                if memory[memptr] > 0:
+                    memory[memptr] = memory[memptr] - 1
+                else:
+                    memory[memptr] = 255
+
+            elif cmd == ">":
+                if memptr == len(memory)-1:
+                    memory.append(0)
+                memptr += 1
+
+            elif cmd == "<":
+                if memptr == 0:
+                    memptr = len(memory) - 1
+                else:
+                    memptr -= 1
+
+            # if we're at the beginning of the loop and the memory is 0,
+            # exit the loop
+            elif cmd == "[" and memory[memptr] == 0:
+                codeptr = bracemap[codeptr]
+
+            # if we're at the end of the loop and the memory is >0, jmp to
+            # the beginning of the loop
+            elif cmd == "]" and memory[memptr]:
+                codeptr = bracemap[codeptr]
+
+            # store the output as a string instead of printing it out
+            elif cmd == ".":
+                result += chr(memory[memptr])
+
+            codeptr += 1
+
+        return result
+
+    def bracemap_and_check(self, program: str) -> Tuple[Optional[Dict], bool]:
+        """
+        Create a bracemap of brackets in the program, to compute jmps.
+        Maps open -> close brackets as well as close -> open brackets.
+
+        Also returns True if the program is legal Brainfuck code. If False, we
+        won't try to even run it.
+        """
+
+        open_stack = []
+        bracemap = dict()
+        legal_instructions = {
+            "+", "-", ">", "<", "[", "]", ".", ",", "\t", "\n", " "
```

Comment:
I see you only count \t and \n as valid characters. Why not just do \s instead and allow all whitespace? You already said that it shouldn't matter if something is all whitespace.

Reply:
Honestly it just didn't cross my mind when I wrote it, definitely the better way to go.